Functor gradient

lindsayad opened the pull request:
@GiudGiud you mentioned trying to do some CRAB modeling this weekend. I don't know if you were planning to use the functor system, and if yes also planning on trying to update @freiler's turbulence work to use it? If so I believe you'll need this add-on for doing functor evaluation of gradients. I'm not going to hold up the other PR with this added work, so I'm opening this new one.

lindsayad committed:
Add gradient() method to FunctorInterface

The need for this became immediately apaprent because of Ramiro's
turbulent viscosity computation which requires gradients

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/817812/) on a61c2b6 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/18806/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Generate and verify coverage](https://civet.inl.gov/job/817841/) on a61c2b6 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>0d650f</th>
      <th colspan=4>#18806 a61c2b</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>81.48%</td>
      <td>81.40%</td>
      <td>-0.08%</td>
      <td>75.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>69323</td>
      <td>69810</td>
      <td>+487</td>
      <td>612</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15757</td>
      <td>15951</td>
      <td>+194</td>
      <td>204</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/18806/coverage/diff/framework.html)

## Modules coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>0d650f</th>
      <th colspan=4>#18806 a61c2b</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>86.95%</td>
      <td>86.96%</td>
      <td>+0.01%</td>
      <td>96.83%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>90680</td>
      <td>90767</td>
      <td>+87</td>
      <td>244</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>13606</td>
      <td>13613</td>
      <td>+7</td>
      <td>8</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/18806/coverage/diff/modules.html)

## Full coverage reports

<details><summary>Reports</summary>

- [framework](https://mooseframework.inl.gov/docs/PRs/18806/coverage/framework)
- [chemical_reactions](https://mooseframework.inl.gov/docs/PRs/18806/coverage/chemical_reactions)
- [combined](https://mooseframework.inl.gov/docs/PRs/18806/coverage/combined)
- [contact](https://mooseframework.inl.gov/docs/PRs/18806/coverage/contact)
- [external_petsc_solver](https://mooseframework.inl.gov/docs/PRs/18806/coverage/external_petsc_solver)
- [fluid_properties](https://mooseframework.inl.gov/docs/PRs/18806/coverage/fluid_properties)
- [fsi](https://mooseframework.inl.gov/docs/PRs/18806/coverage/fsi)
- [functional_expansion_tools](https://mooseframework.inl.gov/docs/PRs/18806/coverage/functional_expansion_tools)
- [geochemistry](https://mooseframework.inl.gov/docs/PRs/18806/coverage/geochemistry)
- [heat_conduction](https://mooseframework.inl.gov/docs/PRs/18806/coverage/heat_conduction)
- [level_set](https://mooseframework.inl.gov/docs/PRs/18806/coverage/level_set)
- [misc](https://mooseframework.inl.gov/docs/PRs/18806/coverage/misc)
- [navier_stokes](https://mooseframework.inl.gov/docs/PRs/18806/coverage/navier_stokes)
- [peridynamics](https://mooseframework.inl.gov/docs/PRs/18806/coverage/peridynamics)
- [phase_field](https://mooseframework.inl.gov/docs/PRs/18806/coverage/phase_field)
- [porous_flow](https://mooseframework.inl.gov/docs/PRs/18806/coverage/porous_flow)
- [ray_tracing](https://mooseframework.inl.gov/docs/PRs/18806/coverage/ray_tracing)
- [rdg](https://mooseframework.inl.gov/docs/PRs/18806/coverage/rdg)
- [richards](https://mooseframework.inl.gov/docs/PRs/18806/coverage/richards)
- [stochastic_tools](https://mooseframework.inl.gov/docs/PRs/18806/coverage/stochastic_tools)
- [tensor_mechanics](https://mooseframework.inl.gov/docs/PRs/18806/coverage/tensor_mechanics)
- [xfem](https://mooseframework.inl.gov/docs/PRs/18806/coverage/xfem)

</details>

This comment will be updated on new commits.

lindsayad commented:
Mac test failure because of issue documented in libmesh/libmesh#3018

GiudGiud commented:
Ah thank you! That would indeed come in handy. As soon as we merge this there are a few discussions asking about matprop gradients btw, self-reminder to update those
@freiler is very close to a merge on the VTB so I'm waiting on that to update his input to wcnsfv. I'm actually looking at SAM-Pgh coupling with INSFV.
Please let me know what's your next step so we dont overlap.

lindsayad commented:
It sounds like you’re covering both of the things we talked about doing for the crab report. Let me know if I can help at any point. 

In the meantime I’ll be shepherding PRs through that are necessary for the user object PR. Mostly the functor PRs and the libmesh update (the latter is needed for the parallel dual number communication) 

(And working on the TA report) 

> On Sep 12, 2021, at 6:52 PM, Guillaume Giudicelli ***@***.***> wrote:
> 
> ﻿
> Ah thank you! That would indeed come in handy. As soon as we merge this there are a few discussions asking about matprop gradients btw, self-reminder to update those
> @freiler is very close to a merge on the VTB so I'm waiting on that to update his input to wcnsfv. I'm actually looking at SAM-Pgh coupling with INSFV.
> Please let me know what's your next step so we dont overlap.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or unsubscribe.
> Triage notifications on the go with GitHub Mobile for iOS or Android.

GiudGiud committed:
Update the mixing length turbulent viscosity material to functor material properties, refs #16809

GiudGiud committed:
Update INSFVSymmetryVelocityBCs to use functor material property for viscosity, refs #16809

GiudGiud committed:
Update wall function BCs to use functor materials, refs #16809

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/820002/) on a5c1826 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/18806/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/18806/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 7df3b095f66fb205a46d4afc6aa4a23e4b394a95`

lindsayad committed:
Add dot to FunctorInterface

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/820411/) on 6e880f5 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/18806/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/18806/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 7df3b095f66fb205a46d4afc6aa4a23e4b394a95`

lindsayad committed:
Demonstrate transient wcnsfv

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/820533/) on ad10f5e wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/18806/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/18806/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 7df3b095f66fb205a46d4afc6aa4a23e4b394a95`

lindsayad committed:
Add high rayleigh number heavy test

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/820764/) on c42bc6d wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/18806/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/18806/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 7df3b095f66fb205a46d4afc6aa4a23e4b394a95`

lindsayad committed:
Clang-format Guillaumes code :-)

GiudGiud committed:
Add limiter on face mu, which is used to interpolate face material properties

GiudGiud committed:
Evaluate symmetry BC viscoity on the element for now

GiudGiud committed:
Add energy time derivative kernel for weakly conmpressible
Refs #16809

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/821315/) on f839102 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/18806/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/18806/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 7b42ecd3a731e49a38c8cb5f154d628e771ac5e4`

GiudGiud committed:
Delete rhoE from the parameters of the GeneralFluidProperties, this is for a PT formulation

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/821347/) on 59c5152 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/18806/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/18806/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 7b42ecd3a731e49a38c8cb5f154d628e771ac5e4`

GiudGiud committed:
Add general functor fluid properties, refs #16809
Keep derivatives commented for now. They are only needed for non AD, so rn only the FE NS, and it does not use or
really need the functor system

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/821379/) on d14543f wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/18806/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/18806/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 7b42ecd3a731e49a38c8cb5f154d628e771ac5e4`

lindsayad committed:
Update failing NS tests

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/821446/) on a4a5fb2 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/18806/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/18806/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 7b42ecd3a731e49a38c8cb5f154d628e771ac5e4`

lindsayad committed:
Address CIVET failures

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/821727/) on 241f5ea wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/18806/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/18806/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 7b42ecd3a731e49a38c8cb5f154d628e771ac5e4`

GiudGiud committed:
Allow for different variable names in GeneralFluidProps and GeneralFunctorFluidProps, refs #18609

GiudGiud committed:
Add scalar diffusion to use functors for the mixing length.
Add time derivative of cp functor to GeneralFluidProperties, refs #18609

GiudGiud committed:
Add mixing length turbulent diffusion kernels for energy, refs #18609

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/822483/) on 38af140 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/18806/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/18806/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format da4fe72e01a3d9059012746105368f24c3e5f416`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/824647/) on fb65dc6 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/18806/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/18806/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format b9a47a04f1981f6c931ea4ae97a940025724adb5`

lindsayad committed:
Restore use of Real Functions in GenericFunctionFunctorMaterial

Use of a limiter *type* in the `Functor` face overloads instead of an
actual limiter object allows us to keep geometric functor calling
arguments agnostic of the return type or type of the `Functor`.
Consequently we can use a Real `Functor` inside the lambda for an
`ADReal` `Functor` and have everything work out dandy

lindsayad committed:
Remove unused variable parameters in GeneralFluidProps

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/824751/) on 720e7be wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/18806/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/18806/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format b9a47a04f1981f6c931ea4ae97a940025724adb5`

lindsayad committed:
Guillaume needs clang-format save hooks some day

moosebuild commented:
Job [Generate and verify coverage](https://civet.inl.gov/job/825260/) on cc8d449 wanted to post the following:


The following coverage requirement(s) failed:
- navier_stokes coverage rate 76.80% is less than the required 77%

GiudGiud committed:
Transition mixing length momentum term to WCNSFV, refs #18609

GiudGiud committed:
Adapt mixing length turbulence model to functor material properties, thanks to functor material property gradients
Refs #16809

lindsayad committed:
Add SingleSidedFaceArg overload

This name is update for debate. AtomicFaceArg is another candidate

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/827235/) on b0af98e wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/18806/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/18806/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format b79d3c2f805a246394c54cd9eca37748c7d3dd5f`

lindsayad committed:
Prefer MooseFunctorName param for functor parameters

moosebuild commented:
Job [Generate and verify coverage](https://civet.inl.gov/job/827807/) on 371f9d7 wanted to post the following:


The following coverage requirement(s) failed:
- navier_stokes coverage rate 76.80% is less than the required 77%

lindsayad committed:
Make square coarse versions of a couple exodiffing tests

And convert the previous exodiff tests to csvdiff tests

moosebuild commented:
Job [Generate and verify coverage](https://civet.inl.gov/job/829744/) on 8f99dba wanted to post the following:


The following coverage requirement(s) failed:
- fluid_properties coverage rate 82.61% is less than the required 84%
- navier_stokes coverage rate 76.80% is less than the required 77%

moosebuild commented:
Job [Generate and verify coverage](https://civet.inl.gov/job/831312/) on 9f22eef wanted to post the following:


The following coverage requirement(s) failed:
- `fluid_properties` coverage rate 82.61% is less than the required 84.0%
- `navier_stokes` coverage rate 76.80% is less than the required 77.0%

lindsayad committed:
Update some heavy tests

lindsayad committed:
Split limiter change: framework/unit/core NS

lindsayad committed:
Split limiter change: new WCNSFV NS

lindsayad committed:
Update for removal of aux from functor prop

lindsayad committed:
Make base evaluateGradient and evaluateDot not implemented

Prefer this over having them pure because I anticipate there are a fair
number of developers who only care about the core evaluate methods and
won't need to touch gradient or dot

lindsayad committed:
Apply suggestions from code review

Co-authored-by: Guillaume Giudicelli <guillaume.giudicelli@gmail.com>

lindsayad committed:
Remove square coarse exodiff versions of larger, now csvdiff tests

The requirement for these tests would be the same, which illuminates the
fact that they probably just shouldn't exist

GiudGiud committed:
Add all the non AD derivatives to the functor fluid props, for compatiblity with non-AD materials
Make findUStar use AD for the wall boundary conditions (no need for auxkernels, since no AD for aux variables)
Use wall viscosity in symmetry BCs

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/851599/) on aa6aa2c wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/18806/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/18806/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 72060375990a6b4756f4964242078db7d1371c82`

GiudGiud committed:
Clang format
Fixup new header for WCNSFV time derivative kernels
Use the temperature gradient instead of the energy graident for turbulent diffusion
Use the singlesidedFaceArg overload to avoid computing ghost gradient for viscosity on symmtrry axis

GiudGiud committed:
Add 2D wcnsfv channel test
Link to wcnsfv in index
Add wcnsfv short design doc
Address test failures: no mixing length on distributed meshes rn

GiudGiud committed:
Add missing doc files
Disable recover testing for WCNSFV channels for now
refs #19126

GiudGiud committed:
Restrict INSFV test to global AD indexing

GiudGiud committed:
Rework functor fluid properties to avoid having a rhoFromPTfunctor as well
Clean up and reorganize functors for readability
Add tests for GeneralFunctorFluid props, refs #16806
Increase rel_dif on wcns material test

lindsayad committed:
Create FE gradient-dot evaluation test

lindsayad committed:
Remove gradient/dot code related to the only not-on-the-fly functor method overload

GiudGiud committed:
Revert "Remove FunctorMatPropElementalAux"

This reverts commit b9a921d30cda53103ee2fcfad2ecef51f5c5cc68.

lindsayad committed:
Remove evaluate code related to the only not-on-the-fly functor method overload

I just don't think that interface falls in line with the spirit of the
functor system, so I'm going to remove it. I think the on-the-fly FE
overloads are plenty good with the local caching

lindsayad committed:
Test errors

GiudGiud committed:
Create AD / non AD verisons of FunctorMatPropElementalAux to accomodate both types of matprops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/854798/) on 25b397b wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/18806/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/18806/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format a9a64a94bbe755ed68c5aef8fd06b17980d89b1d`

lindsayad committed:
Update GeneralFunctorFluidProps to FunctorInterface APIs

lindsayad committed:
Test SingleSidedFaceArg and FunctorMatPropElementalAux

lindsayad committed:
Clear local cache in Function during setup
